SmartTutor/ContentModel/SubscriptionService.cs

```diff
@@ -0,0 +1,97 @@
+﻿using System;
+using System.Collections.Generic;
+using SmartTutor.ContentModel.DTOs;
+using SmartTutor.ContentModel.Exceptions;
+using SmartTutor.ContentModel.Lectures;
+using SmartTutor.ContentModel.Subscriptions;
+using SmartTutor.ContentModel.Subscriptions.Repository;
+
+namespace SmartTutor.ContentModel
+{
+    public class SubscriptionService:ISubscriptionService
+    {
+        private readonly ISubscriptionRepository _subscriptionRepository;
+
+        public SubscriptionService(ISubscriptionRepository subscriptionRepository)
+        {
+            _subscriptionRepository = subscriptionRepository;
+        }
+
+        public void SubscribeTeacher(SubscriptionDto dto)
+        {
+            var teacher = _subscriptionRepository.GetTeacher(dto.TeacherId);
+            
+            if (teacher.GetActiveSubscription() != null) throw new TeacherAlreadySubscribedException(dto.TeacherId.ToString());
+            var individualPlan = _subscriptionRepository.GetIndividualPlan(dto.IndividualPlanId);
+            var individualPlanUsage = new IndividualPlanUsage(individualPlan.Id);
+            individualPlanUsage = _subscriptionRepository.SaveOrUpdatePlanUsage(individualPlanUsage);
+            var end = GetEndDate(dto.NumberOfDays);
+            var subscription = new Subscription(dto.TeacherId,DateTime.Now,end, individualPlanUsage.Id);
+            subscription = _subscriptionRepository.SaveOrUpdateSubscription(subscription);
+            teacher.AddSubscription(subscription);
+            _subscriptionRepository.SaveOrUpdateTeacher(teacher);
+        }
+
+        public bool CanAddLecture(int teacherId)
+        {
+            var planUsage = GetIndividualPlanUsage(teacherId);
+            var individualPlan = _subscriptionRepository.GetIndividualPlan(planUsage.IndividualPlanId);
+            return individualPlan.NumberOfLectures - planUsage.NumberOfLecturesUsed >= 1;
+        }
+
+        public bool CanAddCourse(int teacherId)
+        {
+            var planUsage = GetIndividualPlanUsage(teacherId);
+            var individualPlan = _subscriptionRepository.GetIndividualPlan(planUsage.IndividualPlanId);
+            return individualPlan.NumberOfCourses - planUsage.NumberOfCoursesUsed >= 1;
+        }
+
+        public void IncrementNumberOfLectures(int teacherId)
+        {
+            var teacher = _subscriptionRepository.GetTeacher(teacherId);
```

Comment:
The following code:
```
var teacher = _subscriptionRepository.GetTeacher(teacherId);
var subscription = teacher.GetActiveSubscription();
var planUsage = _subscriptionRepository.GetIndividualPlanUsage(subscription.PlanUsageId);
```
Appears four times in this class. It can be extracted into a separate method. Better yet, the logic can be moved to the _subscriptionRepository, where you can introduce a method GetActiveSubscriptionForTeacher(teacherId).

A good heuristic to integrate into your everyday programming - when you finish a larger portion of code, scroll through it and see if there are any similar pieces of code. Adhere to DRY whenever possible.

Reply:
I made that function in line 70 but forgot to call it.  😰